object_gomap_test.go

```diff
@@ -326,3 +328,142 @@ func TestGoMapUnicode(t *testing.T) {
+		val            *objectGoMapSimple
+		threshold      int
+		expectedMem    uint64
+		expectedNewMem uint64
+		errExpected    error
+	}{
+		{
+			name:      "should account for each key value pair given a non-empty object",
+			threshold: 100,
+			val: &objectGoMapSimple{
+				baseObject: baseObject{
+					val: &Object{runtime: vm},
+				},
+				data: map[string]interface{}{
+					"test0": valueInt(99),
+					"test1": valueInt(99),
+					"test2": valueInt(99),
+					"test3": valueInt(99),
+				},
+			},
+			// overhead + len("testN") + value
+			expectedMem: SizeEmptyStruct + (5+SizeInt)*4,
+			// overhead + len("testN") with string overhead + value
+			expectedNewMem: SizeEmptyStruct + ((5+SizeString)+SizeInt)*4,
+			errExpected:    nil,
+		},
+		{
+			name:      "should account for each key value pair given map with a nil value",
+			threshold: 100,
+			val: &objectGoMapSimple{
+				baseObject: baseObject{
+					val: &Object{runtime: vm},
+				},
+				data: map[string]interface{}{
+					"test": nil,
+				},
+			},
+			// overhead + len("test") + null
+			expectedMem: SizeEmptyStruct + 4 + SizeEmptyStruct,
+			// overhead + len("test") with string overhead + null
+			expectedNewMem: SizeEmptyStruct + (4 + SizeString) + SizeEmptyStruct,
+			errExpected:    nil,
+		},
+		{
+			name:      "should account for nested parsed map",
+			threshold: 100,
+			val: &objectGoMapSimple{
+				baseObject: baseObject{
+					val: &Object{runtime: vm},
+				},
+				data: map[string]interface{}{
+					"test": &objectGoMapSimple{ // <- this is treated as a objectGoMapReflect
+						baseObject: baseObject{
+							val: &Object{runtime: vm},
+						},
+						data: map[string]interface{}{
+							"subTest1": valueInt(99),
+							"subTest2": valueInt(99),
+							"subTest3": valueInt(99),
+							"subTest4": valueInt(99),
+						},
+					},
+				},
+			},
+			// overhead + len("test") + reflectedMap
+			expectedMem: SizeEmptyStruct + 4 + SizeEmptyStruct,
+			// overhead + len("test") with string overhead + reflectedMap
+			expectedNewMem: SizeEmptyStruct + 4 + SizeString + SizeEmptyStruct,
+			errExpected:    nil,
+		},
+		{
+			name:      "should account for nested key value pairs",
+			threshold: 100,
+			val: &objectGoMapSimple{
+				baseObject: baseObject{
+					val: &Object{runtime: vm},
+				},
+				data: map[string]interface{}{
+					"test": map[string]interface{}{ // is this ever valid/should we test this situation?
+						"subTest1": valueInt(99),
+						"subTest2": valueInt(99),
+						"subTest3": valueInt(99),
+						"subTest4": valueInt(99),
+					},
+				},
+			},
+			// overhead + len("test") + VM runtime base object + values
+			expectedMem: SizeEmptyStruct + 4 + 2204 + ((8 + SizeInt) * 4),
+			// overhead + len("testN") with string overhead + VM runtime base object with overhead + values with string overhead
+			expectedNewMem: SizeEmptyStruct + (4 + SizeString) + 4716 + ((8 + SizeString + SizeInt) * 4),
+			errExpected:    nil,
+		},
```

Comment:
[Q] Same as above, is this test case even possible/valid? 

Also just want to point out that when nested there is a significant size difference? Is this because there is like a duplicate vm/runtime instance. I tried following down the debug breakpoint rabbit hole, but definitely gets confusing with all the recursive calls.

Reply:
I think this is a valid test case and I'm looking to see why there's so much additional memory.